### ISSUE TITLE
batchnorm1d inference behavior is incorrect

### ISSUE THREAD

**Issue by geneing** [pytorch/pytorch#16793]
## 🐛 Bug

When in inference mode batchnorm1d should use precomputed weights and means. However, code contains a typo in:
`F.batch_norm(
            input, self.running_mean, self.running_var, self.weight, self.bias,
            self.training or not self.track_running_stats,
            exponential_average_factor, self.eps)`

I believe it should be `self.training or self.track_running_stats`. Currently user has to set training=False, and track_running_stats=True in inference mode for batchnorm to "not" update running mean and variance.

## To Reproduce

If "layer" is a pretrained batchnorm1d layer loaded from saved state:
` ##dummy input x
    x=np.matmul((1.+1./np.arange(1,129))[:,np.newaxis], (-3 + 2./np.arange(1,11))[np.newaxis,:]) 
    xt=torch.tensor(x[np.newaxis,:,:],dtype=torch.float32)
    
    weight=layer.weight.detach().numpy()
    bias=layer.bias.detach().numpy()
    running_mean = layer.running_mean.detach().numpy()
    running_var = layer.running_var.detach().numpy()
    x1=xt.detach().numpy().squeeze()
    eps = layer.eps
    y = ((x1[:,0]-running_mean)/(np.sqrt(running_var+eps)))*weight+bias
    
    layer.track_running_stats=False
    layer.training = False
    c = layer(xt)`

## Expected behavior

I expect "y" to be the same as "c", but it's not unless layer.track_running_stats=True

## Environment

 - PyTorch: 1.0
 - OS: Linux
 - How you installed PyTorch (`conda`, `pip`, source):
 - Build command you used (if compiling from source):
 - Python version: 3.6

## Additional context
Bug is still in the current git master. https://github.com/pytorch/pytorch/blob/255136fc1dd040fd51b346b459d2d8fb8d1f895e/torch/nn/modules/batchnorm.py#L75

**Comment by gchanan** [comment]
@SsnL

**Comment by SsnL** [comment]
No the current behavior is correct. When `self.track_running_stats = True`, the bool expression equals `self.training`, which is what we want. And when `self.track_running_stats = False`, batch stats are always used, because you asked it to not track running stats and it didn't.

**Comment by geneing** [comment]
@SsnL I believe you are mistaken. Look here...

If self.training=False and self.track_running_stats=False, we get:

False or (Not False) == False or True == True. 

Which means that we asked not to track running stats in inference, but batchnorm is tracking and using running stats. 

I think the code should be "self.training or self.track_running_stats". Then when training=True, we always track running stats. 

For inference training=False. Then track_running_stats=True => whole expression true, we are tracking stats. With track_running_stats=False => whole expression is false, we are using mean/variance computed during training.

Better yet - don't overload one parameter with two meanings.

**Comment by SsnL** [comment]
@geneing No. `track_running_stats` means whether the layer tracks running stats in general (if it is true, it uses running stats in eval), not whether this particular call tracks. The doc is very clear on this.

**Comment by LcDog** [comment]
The `running_mean` and `running_var` are still updating, though `track_running_stats = False`. Am I right? (But `running_mean` and `running_var` will not be used in inference.) https://github.com/pytorch/pytorch/blob/d7a1152ee9948e21f337264460c5f6d12e4ed67f/torch/onnx/symbolic_opset9.py#L900-L907

**Comment by jizongFox** [comment]
@SsnL  See the reproduction of the bug:
```
import torch 
bn=torch.nn.BatchNorm2d(3) 
print(bn.running_mean) 
#tensor([0., 0., 0.])
bn.track_running_stats = False 
for i in range (10): 
    bn(torch.rand(1, 3, 2, 2)) 
print(bn.running_mean)
#tensor([0.32, 0.12, 0.12])
```
There is indeed something wrong with this implementation for `PyTorch 1.3.0`

**Comment by SsnL** [comment]
@jizongFox `track_running_stats ` is meat to be set at constructor, not afterwards, because it changes how buffers area created. Also, what you described is unrelated from what this issue was about.
